service/controller/v17/resource/ipam/create.go

```diff
@@ -0,0 +1,167 @@
+	}
+
+	return nil
+}
+
+func (r *Resource) allocateSubnet(ctx context.Context) (string, error) {
+	var reservedSubnets []net.IPNet
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "getting allocated subnets from VPCs")
+		vpcSubnets, err := getVPCSubnets(ctx)
+		if err != nil {
+			return "", microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", "got allocated subnets from VPCs")
+
+		reservedSubnets = append(reservedSubnets, vpcSubnets...)
+	}
+
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "getting allocated subnets from AWSConfigs")
+		awsConfigSubnets, err := getAWSConfigSubnets(r.g8sClient)
+		if err != nil {
+			return "", microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", "got allocated subnets from AWSConfigs")
+
+		reservedSubnets = append(reservedSubnets, awsConfigSubnets...)
+	}
+
+	reservedSubnets = canonicalizeSubnets(r.networkRange, reservedSubnets)
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", "finding free subnet")
+	subnet, err := ipam.Free(r.networkRange, r.allocatedSubnetMask, reservedSubnets)
+	if err != nil {
+		return "", microerror.Maskf(err, "networkRange: %s, allocatedSubnetMask: %s, reservedSubnets: %#v", r.networkRange.String(), r.allocatedSubnetMask.String(), reservedSubnets)
+	}
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found free subnet: %s", subnet.String()))
+
+	return subnet.String(), nil
+}
+
+func canonicalizeSubnets(network net.IPNet, subnets []net.IPNet) []net.IPNet {
+	// Naive deduplication as net.IPNet cannot be used as key for map. This
+	// should be ok for current foreseeable future.
+	for i := 0; i < len(subnets); i++ {
+		// Remove subnets that don't belong to our desired network.
+		if !network.Contains(subnets[i].IP) {
+			subnets = append(subnets[:i], subnets[i+1:]...)
+			i--
+			continue
+		}
+
+		// Remove duplicates.
+		for j := i + 1; j < len(subnets); j++ {
+			if reflect.DeepEqual(subnets[i], subnets[j]) {
+				subnets = append(subnets[:j], subnets[j+1:]...)
+				j--
+			}
+		}
+	}
+
+	return subnets
+}
+
+func getAWSConfigSubnets(g8sClient versioned.Interface) ([]net.IPNet, error) {
+	awsConfigList, err := g8sClient.ProviderV1alpha1().AWSConfigs(apismetav1.NamespaceAll).List(apismetav1.ListOptions{})
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	var results []net.IPNet
+	for _, ac := range awsConfigList.Items {
+		cidr := key.ClusterNetworkCIDR(ac)
+		if cidr == "" {
+			continue
+		}
+
+		_, n, err := net.ParseCIDR(cidr)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+
+		results = append(results, *n)
+	}
+
+	return results, nil
+}
+
+func getVPCSubnets(ctx context.Context) ([]net.IPNet, error) {
+	sc, err := controllercontext.FromContext(ctx)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	subsOut, err := sc.AWSClient.EC2.DescribeSubnets(nil)
```

Comment:
Will this fetch all subnets? What amounts are we talking about? Is there paging support?

Reply:
As far as I can read documentation here https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeSubnets.html I cannot figure a way how it would return partial response. Number of subnets is limited for AWS Account anyway, so it cannot be astronomical number. In average installation, I'd figure this is something like 20-40 subnets total.